File: typescript/134-Gas-Station.ts
function canCompleteCircuit(gas: number[], cost: number[]): number {
    let res = 0;
    let netDistance = 0;

    // Checks if there is enough gas to complete a cycle
    if (gas.reduce((a, b) => a + b) - cost.reduce((a, b) => a + b) < 0) {
        return -1;
    }

    // finds positive netDistance to check if cycle can be completed
    for (let i = 0; i < gas.length; i++) {
        netDistance += gas[i] - cost[i];

        // resets net Distance and gets gas starting gas station
        if (netDistance < 0) {
            netDistance = 0;
            res = i + 1;
        }
    }
    return res;
}

File: swift/134-Gas-Station.swift
class Solution {
    func canCompleteCircuit(_ gas: [Int], _ cost: [Int]) -> Int {
        if gas.reduce(0, +) < cost.reduce(0, +) {
            return -1
        }
        var start = gas.count - 1, end = 0
        var total = gas[start] - cost[start]
        
        while start >= end {
            while total < 0 && start >= end {
                start -= 1
                total += gas[start] - cost[start]
            }
            if start == end {
                return start
            }
            total += gas[end] - cost[end]
            end += 1
        }
        return -1
    }
}